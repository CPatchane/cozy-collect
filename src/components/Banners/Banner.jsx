import React, { PureComponent } from 'react'
import cx from 'classnames'
import PropTypes from 'prop-types'

import Text, { SubTitle } from 'cozy-ui/react/Text'

export class Banner extends PureComponent {
  render() {
    const { actionButton, description, isImportant, title } = this.props
    return (
      <div
        className={cx(
          'u-bdrs-4 u-p-1-half u-bg-paleGrey',
          isImportant ? 'u-bg-chablis' : ''
        )}
      >
        <SubTitle className={isImportant ? 'u-pomegranate' : ''}>
          {title}
        </SubTitle>
        <Text className="u-mv-1">{description}</Text>
        {!!actionButton && actionButton}
      </div>
    )
  }
}

Banner.propTypes = {
  actionButton: PropTypes.element,
  description: PropTypes.string.isRequired,
  isImportant: PropTypes.bool,
  title: PropTypes.string.isRequired
}

export default Banner
