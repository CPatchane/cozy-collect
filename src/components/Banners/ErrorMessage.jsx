import React, { Component } from 'react'
import { translate } from 'cozy-ui/react/I18n'
import { isKonnectorTwoFAError } from 'lib/konnectors'
import Button from 'cozy-ui/react/Button'
import PropTypes from 'prop-types'

import Banner from 'components/Banners/Banner'
import { getMostAccurateErrorKey, isKonnectorKnownError } from 'lib/konnectors'

export class ErrorMessage extends Component {
  constructor(props) {
    super(props)
    this.renderButton = this.renderButton.bind(this)
    this.getErrorTitle = this.getErrorTitle.bind(this)
    this.getErrorDescription = this.getErrorDescription.bind(this)
  }

  renderButton(error) {
    const { t, onForceConnection, disabled } = this.props
    if (isKonnectorTwoFAError(error)) {
      return (
        <Button
          label={t('connection.CTA.twofa_failed')}
          icon="sync"
          theme="danger"
          className="u-m-0"
          onClick={onForceConnection}
          disabled={disabled}
        />
      )
    } else {
      return null
    }
  }

  getErrorTitle() {
    const { t, error } = this.props
    return t(
      getMostAccurateErrorKey(t, error, key => `connection.error.${key}.title`)
    )
  }

  getErrorDescription() {
    const { t, konnector, error } = this.props
    return t(
      getMostAccurateErrorKey(
        t,
        error,
        key => `connection.error.${key}.description`
      ),
      {
        name: konnector.name,
        link: konnector.vendor_link
      }
    )
  }

  render() {
    const { t, konnector, error } = this.props

    if (!isKonnectorKnownError(error)) {
      return (
        <Banner
          description={t('connection.error.default.description', {
            name: konnector.name
          })}
          title={t('connection.error.default.title')}
          isImportant
        />
      )
    }

    // FIXME temporarily, only for EDF
    if (konnector.slug === 'edf') {
      return (
        <Banner
          description={t('status.edf.maintenance', {
            supportLink: t('status.edf.support_link')
          })}
          title={t('status.interrupted')}
          isImportant
        />
      )
    }

    return (
      <Banner
        actionButton={this.renderButton(error)}
        description={this.getErrorDescription(error)}
        isImportant
        title={this.getErrorTitle(error)}
      />
    )
  }
}

ErrorMessage.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.object.isRequired,
  konnector: PropTypes.object.isRequired,
  onForceConnection: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(ErrorMessage)
