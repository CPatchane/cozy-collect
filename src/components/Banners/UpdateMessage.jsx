/* global cozy */

import React, { Component } from 'react'
import { translate } from 'cozy-ui/react/I18n'
import Button from 'cozy-ui/react/Button'
import PropTypes from 'prop-types'

import Banner from 'components/Banners/Banner'

export class UpdateMessage extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isRedirecting: false
    }
    this.redirectToStore = this.redirectToStore.bind(this)
  }

  async redirectToStore() {
    this.setState({ isRedirecting: true })
    const { konnector } = this.props
    try {
      await cozy.client.intents.redirect('io.cozy.apps', {
        slug: konnector.slug,
        step: 'update'
      })
    } catch (error) {
      /* eslint-disable-next-line no-console */
      console.error(error)
      this.setState({ isRedirecting: false })
    }
  }

  render() {
    const { t, isBlocking } = this.props
    const { isRedirecting } = this.state

    return (
      <Banner
        actionButton={
          <Button
            label={t('update.CTA')}
            theme="danger"
            className="u-m-0"
            onClick={this.redirectToStore}
            disabled={isRedirecting}
          />
        }
        description={isBlocking ? t('update.blocking') : t('update.regular')}
        isImportant={isBlocking}
        title={t('update.title')}
      />
    )
  }
}

UpdateMessage.propTypes = {
  konnector: PropTypes.object.isRequired,
  isBlocking: PropTypes.bool,
  t: PropTypes.func.isRequired
}

export default translate()(UpdateMessage)
