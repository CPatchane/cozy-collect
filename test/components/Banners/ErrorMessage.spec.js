'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { tMock } from '../../jestLib/I18n'
import { ErrorMessage } from 'components/Banners/ErrorMessage'

const konnectorMock = {
  slug: 'mocK'
}

const errorMock = {
  type: 'LOGIN_FAILED',
  code: 'LOGIN_FAILED'
}

const twoFaExpiredErrorMock = {
  type: 'USER_ACTION_NEEDED',
  code: 'USER_ACTION_NEEDED.TWOFA_EXPIRED'
}

const twoFaWrongErrorMock = {
  type: 'USER_ACTION_NEEDED',
  code: 'USER_ACTION_NEEDED.WRONG_TWOFA_CODE'
}

const onForceConnectionMock = () => jest.fn()

describe('ErrorMessage component', () => {
  it(`Should render correctly`, () => {
    const component = shallow(
      <ErrorMessage
        konnector={konnectorMock}
        t={tMock}
        error={errorMock}
        onForceConnection={onForceConnectionMock}
      />
    )
      .dive()
      .getElement()
    expect(component).toMatchSnapshot()
  })

  it(`Should render correctly if two session expired`, () => {
    const component = shallow(
      <ErrorMessage
        konnector={konnectorMock}
        t={tMock}
        error={twoFaExpiredErrorMock}
        onForceConnection={onForceConnectionMock}
      />
    )
      .dive()
      .getElement()
    expect(component).toMatchSnapshot()
  })

  it(`Should render correctly if wrong twofa code error`, () => {
    const component = shallow(
      <ErrorMessage
        konnector={konnectorMock}
        t={tMock}
        error={twoFaWrongErrorMock}
        onForceConnection={onForceConnectionMock}
      />
    )
      .dive()
      .getElement()
    expect(component).toMatchSnapshot()
  })
})
