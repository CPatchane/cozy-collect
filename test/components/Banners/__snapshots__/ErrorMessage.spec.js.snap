// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage component Should render correctly 1`] = `
<div
  className="u-flex u-flex-column u-maw-6 u-mih-2 u-bdrs-4 u-ph-1 u-pt-1 u-bg-paleGrey u-ta-left u-bg-chablis u-maw-none"
>
  <SubTitle
    className="u-mb-1 u-pomegranate"
    ellipsis={false}
    tag="div"
  >
    Bad credentials
  </SubTitle>
  <div
    className="u-flex u-w-100 u-mb-1"
  >
    <Text
      className="u-ta-left"
      ellipsis={false}
      tag="div"
    >
      Bad credentials. Check the konnector fields and run the connection again.
    </Text>
  </div>
</div>
`;

exports[`ErrorMessage component Should render correctly if two fa session expired 1`] = `
<div
  className="u-flex u-flex-column u-maw-6 u-mih-2 u-bdrs-4 u-ph-1 u-pt-1 u-bg-paleGrey u-ta-left u-bg-chablis u-maw-none"
>
  <SubTitle
    className="u-mb-1 u-pomegranate"
    ellipsis={false}
    tag="div"
  >
    Relaunch the connexion to the service to fetch your data
  </SubTitle>
  <div
    className="u-flex u-w-100 u-mb-1"
  >
    <Text
      className="u-ta-left"
      ellipsis={false}
      tag="div"
    >
      The last connexion to the service failed; please launch it again. You may have to provide a validation code.
    </Text>
  </div>
  <div
    className="u-flex-shrink-0 u-mb-1"
  >
    <ComponentWithLocales
      trigger={Object {}}
    >
      [Function]
    </ComponentWithLocales>
  </div>
</div>
`;

exports[`ErrorMessage component Should render correctly if wrong twofa code error 1`] = `
<div
  className="u-flex u-flex-column u-maw-6 u-mih-2 u-bdrs-4 u-ph-1 u-pt-1 u-bg-paleGrey u-ta-left u-bg-chablis u-maw-none"
>
  <SubTitle
    className="u-mb-1 u-pomegranate"
    ellipsis={false}
    tag="div"
  >
    Wrong two-factore code provided
  </SubTitle>
  <div
    className="u-flex u-w-100 u-mb-1"
  >
    <Text
      className="u-ta-left"
      ellipsis={false}
      tag="div"
    >
      The two-factor code provided is wrong, please start again.
    </Text>
  </div>
  <div
    className="u-flex-shrink-0 u-mb-1"
  >
    <ComponentWithLocales
      trigger={Object {}}
    >
      [Function]
    </ComponentWithLocales>
  </div>
</div>
`;
